storage: use latest supported table format when writing SSTs

RaduBerinde opened the pull request:
Fixes #151085

#### roachpv: add Version.Cmp method

Add a `Version.Cmp()` method that is more natural to use than
`Less/LessEq/AtLeast` and can also be used for things like
`slices.Sort`.

Epic: none
Release note: None

#### storage: clarify pebbleFormatVersion and add a test

Epic: none
Release note: None

#### storage: use latest supported table format when writing SSTs

When CRDB is writing SSTs (for ingestion or transport), it is using a
fixed table format version. We improve this to use the latest version
guaranteed to be supported by all nodes in the cluster (based on the
current cluster version).

During upgrade, the Pebble format is bumped at fence versions (e.g.
24.3-7) which guarantees that all nodes are using that version at the
non-fence version (e.g. 24.3-8).

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151176)
<!-- Reviewable:end -->

annrpom reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-:-OWXTwfk6-Jy2sIDj1dc:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens)



<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
I think @sumeerbhola brought up a concern a few months back that this is not sufficient. When cluster version _A_ becomes active, it causes two things to happen:
1) the DB ratchets the format major version, beginning to accept ingest's for a new table format
2) SSTs sent over the wire start being created with the latest FMV's max table format

We need 1) to always precede 2), and I'm not sure we have the guarantee right now

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-:-OWXfX0TC_JJtUaSUlSE:bdifr7b)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I see. I think I have a solution. If our node is at version `v`, then it should be guaranteed that all other nodes are at version at least `v-1`. This is because we bump the version on all nodes one step at a time. So it would be safe for these methods to use `pebbleFormatVersion(v-1)`. To ensure we use the latest version, we can just add an internal version (gate) and associate *that* with the new major version (instead of associating the final V25_4 release). WDYT?

sumeerbhola reviewed (changes requested):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r3, 2 of 2 files at r4, 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-:-OWqocG68IrQJ3aWhZmI:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[`pkg/storage/pebble.go` line 2343 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqjERv30Cu7jRn9mgT:-OWqjERv30Cu7jRn9mgU:biv0du7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a3ade33fd027212782281316ddd0d0748ec4bfd/pkg/storage/pebble.go#L2343)):*
> ```Go
> // default created with this version. It should correspond to the minimum
> // supported binary version.
> const MinimumSupportedFormatVersion = pebble.FormatTableFormatV6
> ```

Did this first commit duplicate this field, so won't compile?
___
*[`pkg/storage/pebble.go` line 2375 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqo4Dv8_monZfneMOj:-OWqo4Dv8_monZfneMOk:bek2el0) ([raw file](https://github.com/cockroachdb/cockroach/blob/935a30c7edde96c7747cbf1dc8a5e0b779884cf1/pkg/storage/pebble.go#L2375)):*
> ```Go
> 	for _, k := range pebbleFormatVersionKeys {
> 		// We switch to using a new format as soon as we reach the fence version.
> 		// Note that at this point, the cluster might contain a node with an older
> ```

same question/comment about older binary. We care about the cluster as a whole because one node can send files to another, and the phrasing of this seems to contradict that.
___
*[`pkg/storage/pebble.go` line 2352 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqkjhnAZtRKmwAArNq:-OWqkjhnAZtRKmwAArNr:b-6124cm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2df105adbe97c299f7d81d3fae49ba4dcc4861c3/pkg/storage/pebble.go#L2352)):*
> ```Go
> 		// We switch to using a new format as soon as we reach the fence version.
> 		// Note that at this point, the cluster might contain a node with an older
> 		// binary; but this node's local Pebble format should not affect other nodes.
> ```

I don't quite understand this comment regarding "older binary".
Say v5 is a fence version and v6 is the user specified version. When the cluster reaches v5, there should be no binary in that cluster that can't run the code for v5 or v6. Of course, that binary may not yet know that the cluster version is v5 (since it is lagging wrt that information). At that point we update the format major version of Pebble corresponding to v6.

Does the migration system run migrations at fence versions? And is it guaranteeing that before it informs any node about the cluster version stepping to v6, it has run the v5 migrations at all nodes, even though v5 is a fence version (i.e., is it doing exactly what we would expect it to do for normal versions)?

(unrelated to this PR) The format major version bumps for Pebble (in a running node) are happening from `kvstorage.WriteClusterVersionToEngines`, which is being called in `bumpClusterVersion`. I've forgotten the details of the migration framework, but want to confirm that `MigrationServer.BumpClusterVersion` is the path that ensures migrations are run, so the fact that we are hooking into this synchronous path for the Pebble format major version bumps is sufficient for correctness.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-:-OWqtP1V3M6Ay1N87tii:b-814g8k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola)

___
*[`pkg/storage/pebble.go` line 2343 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqjERv30Cu7jRn9mgT:-OWqpsdMDhZ9Il24No9O:blk0w72) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a3ade33fd027212782281316ddd0d0748ec4bfd/pkg/storage/pebble.go#L2343)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Did this first commit duplicate this field, so won't compile?
</blockquote></details>

I just moved it.
___
*[`pkg/storage/pebble.go` line 2352 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqkjhnAZtRKmwAArNq:-OWqq3Nr8t5Qn7YZG-Bd:bvdjn7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2df105adbe97c299f7d81d3fae49ba4dcc4861c3/pkg/storage/pebble.go#L2352)):*
Cleaned up the comment.

> The format major version bumps for Pebble (in a running node) are happening from `kvstorage.WriteClusterVersionToEngines`, which is being called in `bumpClusterVersion`

Correct. Note that there are two `bumpClusterVersion`s: `upgrademanager.bumpClusterVersion()` is the "client side" and `server.bumpClusterVersion()` is the "server side".

> Does the migration system run migrations at fence versions?

No migrations, but what we care about is that it does `upgrademanager.bumpClusterVersion()`: https://github.com/cockroachdb/cockroach/blob/3f19e51dcaa5a0a555757acedaf3eacd24fedeae/pkg/upgrade/upgrademanager/manager.go#L476

> And is it guaranteeing that before it informs any node about the cluster version stepping to v6, it has run the v5 migrations at all nodes, even though v5 is a fence version (i.e., is it doing exactly what we would expect it to do for normal versions)?

Yes, `upgrademanager.bumpClusterVersion()` guarantees this, through `forEveryNodeUntilClusterStable`:
https://github.com/cockroachdb/cockroach/blob/3f19e51dcaa5a0a555757acedaf3eacd24fedeae/pkg/upgrade/upgrademanager/manager.go#L588
___
*[`pkg/storage/pebble.go` line 2375 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqo4Dv8_monZfneMOj:-OWqtMLNAQnGeChpcXQj:b-dwiyr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/935a30c7edde96c7747cbf1dc8a5e0b779884cf1/pkg/storage/pebble.go#L2375)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

same question/comment about older binary. We care about the cluster as a whole because one node can send files to another, and the phrasing of this seems to contradict that.
</blockquote></details>

Fixed the comment.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-:-OWuY2f_4nzGFKSrd7Gu:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[`pkg/storage/pebble.go` line 2352 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqkjhnAZtRKmwAArNq:-OWuXi_I2qlzHjxSsv0Q:b-vxmen) ([raw file](https://github.com/cockroachdb/cockroach/blob/2df105adbe97c299f7d81d3fae49ba4dcc4861c3/pkg/storage/pebble.go#L2352)):*
> No migrations, but what we care about is that it does `upgrademanager.bumpClusterVersion()`

Could you add these details in a code comment. Otherwise this will come up later again wrt how is this correct given migrations don't run on fence versions.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola)

___
*[`pkg/storage/pebble.go` line 2352 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWqkjhnAZtRKmwAArNq:-OWusdzF47ZXJpC-3-MA:b-qqx0w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2df105adbe97c299f7d81d3fae49ba4dcc4861c3/pkg/storage/pebble.go#L2352)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> No migrations, but what we care about is that it does `upgrademanager.bumpClusterVersion()`

Could you add these details in a code comment. Otherwise this will come up later again wrt how is this correct given migrations don't run on fence versions.
</blockquote></details>

Done. TFTR!


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachpv: add Version.Cmp method

Add a `Version.Cmp()` method that is more natural to use than
`Less/LessEq/AtLeast` and can also be used for things like
`slices.Sort`.

Epic: none
Release note: None

RaduBerinde committed:
storage: clarify pebbleFormatVersion and add a test

Epic: none
Release note: None

RaduBerinde committed:
storage: use latest supported table format when writing SSTs

When CRDB is writing SSTs (for ingestion or transport), it is using a
fixed table format version. We improve this to use the latest version
guaranteed to be supported by all nodes in the cluster (based on the
current cluster version).

During upgrade, the Pebble format is bumped at fence versions (e.g.
24.3-7) which guarantees that all nodes are using that version at the
non-fence version (e.g. 24.3-8).

Epic: none
Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-:-OWwXGb26QGQmH2reafb:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom)

___
*[`pkg/storage/pebble.go` line 2387 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWwX4_0DI-39VpbQ56Y:-OWwX4_0DI-39VpbQ56Z:br0v0k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2acc18ee3b7f07f9356360c7e2a074272f08b78/pkg/storage/pebble.go#L2387)):*
> ```Go
> 		//
> 		// Say that pebbleFormatVersionMap defines a new Pebble format version 19 at
> 		// cluster version 24.3-6. Then pebbleFormatVersion() returns 31 for the
> ```

returns 19?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151176)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @annrpom and @sumeerbhola)

___
*[`pkg/storage/pebble.go` line 2387 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/151176#-OWwX4_0DI-39VpbQ56Y:-OWzFqOJEu0onJsHsaDc:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2acc18ee3b7f07f9356360c7e2a074272f08b78/pkg/storage/pebble.go#L2387)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

returns 19?
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
